File: ios/Classes/Models/Mpgs/MpgsGatewayInfo.swift
//
//  MpgsGatewayInfo.swift
//  trywithmpgs
//
//  Created by Ray Chow on 21/4/2022.
//

import Foundation

struct MpgsGatewayInfo: Mappable {
    let region: MethodChannelMpgsGatewayRegion
    let merchantId: String
    let sessionId: String
    let apiVersion: String
    
    
    static func mapFromDict(dict: [String: Any]) -> MpgsGatewayInfo? {
        guard let region = dict[MethodChannelMpgsGatewayInfoKey.gatewayRegion.rawValue] as? String,
              let merchantId = dict[MethodChannelMpgsGatewayInfoKey.merchantId.rawValue] as? String,
              let sessionId = dict[MethodChannelMpgsGatewayInfoKey.sessionId.rawValue] as? String,
              let apiVersion = dict[MethodChannelMpgsGatewayInfoKey.apiVerions.rawValue] as? String
        else { return nil }
        //TODO: the default value may need to change
        return MpgsGatewayInfo(region: MethodChannelMpgsGatewayRegion(rawValue: region) ?? MethodChannelMpgsGatewayRegion.asiaPacific, merchantId: merchantId, sessionId: sessionId, apiVersion: apiVersion)
    }
}

File: android/src/main/kotlin/com/keychainpay/keychainpay_mpgs/models/mpgs/MpgsGatewayInfo.kt
package com.keychainpay.keychainpay_mpgs.models.mpgs

import com.keychainpay.keychainpay_mpgs.enums.method_channel.card_merchant.mpgs.MethodChannelMpgsGatewayInfoKey
import com.keychainpay.keychainpay_mpgs.enums.method_channel.card_merchant.mpgs.MethodChannelMpgsGatewayRegion
import com.keychainpay.keychainpay_mpgs.models.base.Mappable
import com.keychainpay.keychainpay_mpgs.utils.safeLet

class MpgsGatewayInfo constructor(regionString: String,
                                  val merchantId: String,
                                  val sessionId: String,
                                  val apiVersion: String) {
    val region: MethodChannelMpgsGatewayRegion = MethodChannelMpgsGatewayRegion.values().first { it.key == regionString }
    companion object: Mappable<MpgsGatewayInfo> {
        override fun mapFromHashMap(hashMap: HashMap<String, Any>): MpgsGatewayInfo? {
            return safeLet(
                hashMap[MethodChannelMpgsGatewayInfoKey.GATEWAY_REGION.key] as? String,
                hashMap[MethodChannelMpgsGatewayInfoKey.MERCHANT_ID.key] as? String,
                hashMap[MethodChannelMpgsGatewayInfoKey.SESSION_ID.key] as? String,
                hashMap[MethodChannelMpgsGatewayInfoKey.API_VERSION.key] as? String
            ) { regionString, merchantId, sessionId, apiVersion ->
                MpgsGatewayInfo(regionString, merchantId, sessionId, apiVersion)
            }
        }
    }
}